cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,231 @@
+
+    It requires the Windows drive mounting feature to be enabled and the
+    disk drive must exist for this to succeed.
+
+    Example:
+    # Assuming Windows drives are mounted under /mnt/ and "S:" doesn't exist:
+    p = winpath2wsl("C:\\ProgramData") # p == "/mnt/c/ProgramData/"
+    n = winpath2wsl("S:\\CoolFolder") # n is None (S doesn't exist)
+
+    :param path: string representing a Windows path. The root drive must exist,
+    although the path is not required to.
+    """
+    out, err = subp.subp([WSLPATH_CMD, "-au", path], rcs=[0, 1])
+    if err:
+        LOG.debug(err)
+        return None
+
+    return PurePath(out.rstrip())
+
+
+def cmd_executable() -> Optional[PurePath]:
+    """
+    Returns the Linux path to the Windows host's cmd.exe.
+    """
+
+    mounts = mounted_win_drives()
+    if not mounts:
+        LOG.error("Windows drives are not mounted.")
+        return None
+
+    # cmd.exe path is being stable for decades.
+    candidate = "%s/Windows/System32/cmd.exe"
+    for mnt in mounts:
+        cmd = candidate % mnt
+        if not os.access(cmd, os.X_OK):
+            continue
+
+        LOG.debug("Found cmd.exe at <%s>", cmd)
+        return PurePath(cmd)
+
+    LOG.error("Couldn't find cmd.exe in any mount point.")
+    return None
+
+
+def win_user_profile_dir() -> Optional[PurePath]:
+    """
+    Returns the Windows user profile directory translated as a Linux path
+    accessible inside the current WSL instance.
+    """
+    cmd = cmd_executable()
+    if cmd is None:
+        return None
+
+    # cloud-init runs too early to rely on binfmt to execute Windows binaries.
+    # But we know that `/init` is the interpreter, so we can run it directly.
+    # See /proc/sys/fs/binfmt_misc/WSLInterop[-late]
+    # inside any WSL instance for more details.
+    home, _ = subp.subp(["/init", cmd.as_posix(), "/C", "echo %USERPROFILE%"])
+    home = home.rstrip()
+    if not home:
+        LOG.error("No output from cmd to show the user profile dir.")
+        return None
+
+    return win_path_2_wsl(home.rstrip())
+
+
+def machine_id():
+    """
+    Returns the local machine ID value from /etc/machine-id.
+    """
+    MACHINE_ID_FILE = "/etc/machine-id"
+
+    if util.wait_for_files([MACHINE_ID_FILE], 2.0):
+        LOG.debug("%s file not found", MACHINE_ID_FILE)
+        return None
+
+    return util.load_file(MACHINE_ID_FILE, decode=True)
+
+
+def candidate_user_data_file_names(instance_name) -> List[str]:
+    """
+    Return a list of candidate file names that may contain user-data
+    in some supported format, ordered by precedence.
+    """
+    lsb_rel = util.lsb_release()
+    distribution_id = lsb_rel["id"]
+    release_codename = lsb_rel["codename"]
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, release_codename),
```

Comment:
Could there be an issue with the "codename" value?

On Alpine Linux (with package lsb-release-minimal installed) "lsb_release -c" returns:

```
Codename:       n/a
```

i.e. there is a "/" in the value returned...

Reply:
> The '/' character by itself won't be a problem

I was thinking of whether the "/" might be interpreted as a file path separator.

Reply:
It will certainly be, but that leads to a file not found, which won't be a catastrophic failure. That composes a candidate filename, not exactly a required file, but rather, one possible file.

The absence of a specific codename would restrict Alpine to either use the instance specific user data file or the distro generic `Allpine-all.user-data` or the most generic `config.user-data`.

If that's an issue, we could use the release numbers instead of the codename, which is more inclusive to the distros, but a little bit more cryptic for the end user, IMHO.

Reply:
@dermotbradley makes a good point - previously `util.lsb_release()` was only used in `cc_apt_configure`, so we'll want to make sure that this is tested on other distros. 

Additionally, since other distros might not want to add this new dependency (it isn't installed on centos-stream or alpine by default, probably others too), it probably makes more sense to shift this over to the distro python modules over in cloudinit/distro/__init__.py so that distros can customize as they see fit.

Reply:
@CarlosNihelton - This is what I have in mind - it doesn't fix the wsl unittests, but updates the rest of them:
```diff
diff --git a/cloudinit/config/cc_apt_configure.py b/cloudinit/config/cc_apt_configure.py
index 8a01c54c5..465ee8a14 100644
--- a/cloudinit/config/cc_apt_configure.py
+++ b/cloudinit/config/cc_apt_configure.py
@@ -228,7 +228,7 @@ def _should_configure_on_empty_apt():
     return True, "Apt is available."
 
 
-def apply_apt(cfg, cloud):
+def apply_apt(cfg: dict, cloud: Cloud):
     # cfg is the 'apt' top level dictionary already in 'v3' format.
     if not cfg:
         should_config, msg = _should_configure_on_empty_apt()
@@ -238,7 +238,7 @@ def apply_apt(cfg, cloud):
 
     LOG.debug("handling apt config: %s", cfg)
 
-    release = util.lsb_release()["codename"]
+    release = cloud.distro.lsb_release()["codename"]
     arch = util.get_dpkg_architecture()
     mirrors = find_apt_mirror_info(cfg, cloud, arch=arch)
     LOG.debug("Apt Mirror info: %s", mirrors)
diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index c857a8737..f16f165dc 100644
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -1248,6 +1248,9 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
             "/bin/true",
         ] + (["-cf", config_file, interface] if config_file else [interface])
 
+    def lsb_release(self) -> dict:
+        return lsb_release()
+
 
 def _apply_hostname_transformations_to_url(url: str, transformations: list):
     """
@@ -1449,3 +1452,32 @@ def uses_systemd():
         return stat.S_ISDIR(res.st_mode)
     except Exception:
         return False
+
+
+def lsb_release() -> dict:
+    fmap = {
+        "Codename": "codename",
+        "Description": "description",
+        "Distributor ID": "id",
+        "Release": "release",
+    }
+
+    data = {}
+    try:
+        out = subp.subp(["lsb_release", "--all"], capture=True)
+        for line in out.stdout.splitlines():
+            fname, _, val = line.partition(":")
+            if fname in fmap:
+                data[fmap[fname]] = val.strip()
+        missing = [k for k in fmap.values() if k not in data]
+        if len(missing):
+            LOG.warning(
+                "Missing fields in lsb_release --all output: %s",
+                ",".join(missing),
+            )
+
+    except subp.ProcessExecutionError as err:
+        LOG.warning("Unable to get lsb_release --all: %s", err)
+        data = dict((v, "UNAVAILABLE") for v in fmap.values())
+
+    return data
diff --git a/cloudinit/sources/DataSourceWSL.py b/cloudinit/sources/DataSourceWSL.py
index 0ed51c3af..100c9e5b4 100644
--- a/cloudinit/sources/DataSourceWSL.py
+++ b/cloudinit/sources/DataSourceWSL.py
@@ -140,27 +140,6 @@ def machine_id():
     return util.load_file(MACHINE_ID_FILE, decode=True)
 
 
-def candidate_user_data_file_names(instance_name) -> List[str]:
-    """
-    Return a list of candidate file names that may contain user-data
-    in some supported format, ordered by precedence.
-    """
-    lsb_rel = util.lsb_release()
-    distribution_id = lsb_rel["id"]
-    release_codename = lsb_rel["codename"]
-
-    return [
-        # WSL instance specific:
-        "%s.user-data" % instance_name,
-        # release codename specific
-        "%s-%s.user-data" % (distribution_id, release_codename),
-        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
-        "%s-all.user-data" % distribution_id,
-        # generic, valid for all WSL distros and instances.
-        "config.user-data",
-    ]
-
-
 class DataSourceWSL(sources.DataSource):
     dsname = "WSL"
 
@@ -191,7 +170,9 @@ class DataSourceWSL(sources.DataSource):
         # Notice that file name casing is irrelevant here. Windows filenames
         # are case insensitive. Even though accessed through Linux, path
         # translation just works with whichever casing we try.
-        for filename in candidate_user_data_file_names(self.instance_name):
+        for filename in self.candidate_user_data_file_names(
+            self.instance_name
+        ):
             file = os.path.join(seed_dir, filename)
             if os.path.isfile(file):
                 return PurePath(file)
@@ -222,6 +203,26 @@ class DataSourceWSL(sources.DataSource):
 
         return True
 
+    def candidate_user_data_file_names(self, instance_name) -> List[str]:
+        """
+        Return a list of candidate file names that may contain user-data
+        in some supported format, ordered by precedence.
+        """
+        lsb_rel = self.distro.lsb_release()
+        distribution_id = lsb_rel["id"]
+        release_codename = lsb_rel["codename"]
+
+        return [
+            # WSL instance specific:
+            "%s.user-data" % instance_name,
+            # release codename specific
+            "%s-%s.user-data" % (distribution_id, release_codename),
+            # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+            "%s-all.user-data" % distribution_id,
+            # generic, valid for all WSL distros and instances.
+            "config.user-data",
+        ]
+
 
 # Used to match classes to dependencies
 datasources = [
diff --git a/cloudinit/util.py b/cloudinit/util.py
index 6401a196f..2838e4a29 100644
--- a/cloudinit/util.py
+++ b/cloudinit/util.py
@@ -94,36 +94,6 @@ def get_dpkg_architecture():
     return out.stdout.strip()
 
 
-@lru_cache()
-def lsb_release():
-    fmap = {
-        "Codename": "codename",
-        "Description": "description",
-        "Distributor ID": "id",
-        "Release": "release",
-    }
-
-    data = {}
-    try:
-        out = subp.subp(["lsb_release", "--all"], capture=True)
-        for line in out.stdout.splitlines():
-            fname, _, val = line.partition(":")
-            if fname in fmap:
-                data[fmap[fname]] = val.strip()
-        missing = [k for k in fmap.values() if k not in data]
-        if len(missing):
-            LOG.warning(
-                "Missing fields in lsb_release --all output: %s",
-                ",".join(missing),
-            )
-
-    except subp.ProcessExecutionError as err:
-        LOG.warning("Unable to get lsb_release --all: %s", err)
-        data = dict((v, "UNAVAILABLE") for v in fmap.values())
-
-    return data
-
-
 def decode_binary(blob, encoding="utf-8"):
     # Converts a binary type into a text type using given encoding.
     if isinstance(blob, str):
diff --git a/tests/unittests/config/test_apt_configure_sources_list_v1.py b/tests/unittests/config/test_apt_configure_sources_list_v1.py
index 8772625f2..2a98cae69 100644
--- a/tests/unittests/config/test_apt_configure_sources_list_v1.py
+++ b/tests/unittests/config/test_apt_configure_sources_list_v1.py
@@ -89,7 +89,7 @@ class TestAptSourceConfigSourceList:
             subp, "subp", return_value=("PPID   PID", "")
         )
         mocker.patch("cloudinit.config.cc_apt_configure._ensure_dependencies")
-        lsb = mocker.patch("cloudinit.util.lsb_release")
+        lsb = mocker.patch("cloudinit.distros.lsb_release")
         lsb.return_value = {"codename": "fakerelease"}
         m_arch = mocker.patch("cloudinit.util.get_dpkg_architecture")
         m_arch.return_value = "amd64"
diff --git a/tests/unittests/config/test_apt_configure_sources_list_v3.py b/tests/unittests/config/test_apt_configure_sources_list_v3.py
index 260c8a647..65f6b6321 100644
--- a/tests/unittests/config/test_apt_configure_sources_list_v3.py
+++ b/tests/unittests/config/test_apt_configure_sources_list_v3.py
@@ -160,7 +160,7 @@ class TestAptSourceConfigSourceList:
             "subp",
             return_value=("PPID   PID", ""),
         )
-        lsb = mocker.patch("cloudinit.util.lsb_release")
+        lsb = mocker.patch("cloudinit.distros.lsb_release")
         lsb.return_value = {"codename": "fakerel"}
         m_arch = mocker.patch("cloudinit.util.get_dpkg_architecture")
         m_arch.return_value = "amd64"
diff --git a/tests/unittests/config/test_apt_source_v1.py b/tests/unittests/config/test_apt_source_v1.py
index cb0503ace..a1f858c20 100644
--- a/tests/unittests/config/test_apt_source_v1.py
+++ b/tests/unittests/config/test_apt_source_v1.py
@@ -70,7 +70,7 @@ class TestAptSourceConfig:
     @pytest.fixture(autouse=True)
     def common_mocks(self, mocker):
         mocker.patch(
-            "cloudinit.util.lsb_release",
+            "cloudinit.distros.lsb_release",
             return_value={"codename": self.release},
         )
         mocker.patch(
diff --git a/tests/unittests/config/test_apt_source_v3.py b/tests/unittests/config/test_apt_source_v3.py
index e94bc8b8e..eb7c6d0a4 100644
--- a/tests/unittests/config/test_apt_source_v3.py
+++ b/tests/unittests/config/test_apt_source_v3.py
@@ -53,7 +53,7 @@ class TestAptSourceConfig:
     @pytest.fixture(autouse=True)
     def setup(self, mocker, tmpdir):
         mocker.patch(
-            f"{M_PATH}util.lsb_release",
+            "cloudinit.distros.lsb_release",
             return_value=MOCK_LSB_RELEASE_DATA.copy(),
         )
         mocker.patch(f"{M_PATH}_ensure_dependencies")
```

Reply:
What if we just replace `util.lsb_release()` by `util.get_linux_distro()` (https://github.com/canonical/cloud-init/blob/main/cloudinit/util.py#L582) ?
That function is more distro-agnostic, doesn't imply dependencies and provides all the information this datasource needs.

Reply:
That should work.

Reply:
Let's use the version numbers instead. The Alpine example from above would look too cryptic IMHO: `Alpine-Alpine_Linux_v3.19.user-data` (Rocky seems OK: `Rocky-Blue_Onyx.user-data`).